mixer/adapter/signalfx/tracing.go

```diff
@@ -0,0 +1,192 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package signalfx
+
+import (
+	"context"
+	"net"
+	"strconv"
+	"time"
+
+	"github.com/signalfx/golib/errors"
+	"github.com/signalfx/golib/sfxclient"
+	"github.com/signalfx/golib/trace"
+
+	"istio.io/istio/mixer/pkg/adapter"
+	"istio.io/istio/mixer/template/tracespan"
+)
+
+type tracinghandler struct {
+	ctx        context.Context
+	env        adapter.Env
+	sink       *sfxclient.HTTPSink
+	tracesChan chan []*tracespan.Instance
+}
+
+func (th *tracinghandler) InitTracing(bufferLen int) error {
+	th.tracesChan = make(chan []*tracespan.Instance, bufferLen)
+
+	th.env.ScheduleDaemon(th.sendTraces)
+	return nil
+}
+
+// Pull spans out of tracesChan as they come in and send them to SignalFx.
+func (th *tracinghandler) sendTraces() {
+	for {
+		select {
+		case <-th.ctx.Done():
+			return
+		case insts := <-th.tracesChan:
+			allInsts := th.drainChannel(insts)
+
+			spans := make([]*trace.Span, 0, len(allInsts))
+			for i := range allInsts {
+				span := convertInstance(allInsts[i])
+				if span.ID == "" {
+					continue
+				}
+				spans = append(spans, span)
+			}
+
+			ctx, cancel := context.WithTimeout(th.ctx, 8*time.Second)
+			err := th.sink.AddSpans(ctx, spans)
+			cancel()
+			if err != nil {
+				_ = th.env.Logger().Errorf("Could not send spans to SignalFx: %s", err.Error())
+			}
+		}
+	}
+}
+
+// Pull spans out of the tracesChan until it is empty.  This helps cut down on
+// the number of requests to ingest by batching things when traces come in
+// fast.
+func (th *tracinghandler) drainChannel(initial []*tracespan.Instance) (out []*tracespan.Instance) {
+	out = append(out, initial...)
+	for {
+		select {
+		case insts := <-th.tracesChan:
+			out = append(out, insts...)
+		default:
+			return out
+		}
+	}
+}
+
+// metric.Handler#HandleTraceSpan
+func (th *tracinghandler) HandleTraceSpan(ctx context.Context, values []*tracespan.Instance) error {
+	if th == nil {
+		return nil
+	}
+
+	select {
+	case th.tracesChan <- values:
+		return nil
+	default:
+		return errors.New("dropping span because trace buffer is full -- you should increase the capacity of it")
+	}
+}
+
+var (
+	clientKind = "CLIENT"
+	serverKind = "SERVER"
+)
+
+// Converts an istio span to a SignalFx span, which is currently equivalent to
+// a Zipkin V2 span.
+func convertInstance(istioSpan *tracespan.Instance) *trace.Span {
+	startTime := float64(istioSpan.StartTime.UnixNano()) / 1000
+	endTime := float64(istioSpan.EndTime.UnixNano()) / 1000
+	duration := endTime - startTime
+
+	kind := &serverKind
+	// ClientSpan doesn't seem to populate reliably yet so fall back on a tag
+	if isLocal, ok := istioSpan.SpanTags["context.reporter.local"].(bool); istioSpan.ClientSpan || (ok && !isLocal) {
+		kind = &clientKind
+	}
+
+	// Get rid of this tag since it is no longer needed
+	delete(istioSpan.SpanTags, "context.reporter.local")
```

Comment:
You're mutating the input parameter. I'm not sure anymore whether this is allowed/desirable.

@ozevren Do you recall if we send the same instance objects to multiple adapters?

Reply:
ok, I'll just add a conditional to filter it out when adding span tags to avoid the need to mutate the input.